Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp

```diff
@@ -4931,6 +4932,201 @@ RefPtr<StyleValue> Parser::parse_content_value(Vector<ComponentValue> const& com
+            case CSS::ValueID::TableRow:
+                return CSS::Display { CSS::Display::Internal::TableRow };
+            case CSS::ValueID::TableCell:
+                return CSS::Display { CSS::Display::Internal::TableCell };
+            case CSS::ValueID::TableColumnGroup:
+                return CSS::Display { CSS::Display::Internal::TableColumnGroup };
+            case CSS::ValueID::TableColumn:
+                return CSS::Display { CSS::Display::Internal::TableColumn };
+            case CSS::ValueID::TableCaption:
+                return CSS::Display { CSS::Display::Internal::TableCaption };
+            case CSS::ValueID::RubyBase:
+                return CSS::Display { CSS::Display::Internal::RubyBase };
+            case CSS::ValueID::RubyText:
+                return CSS::Display { CSS::Display::Internal::RubyText };
+            case CSS::ValueID::RubyBaseContainer:
+                return CSS::Display { CSS::Display::Internal::RubyBaseContainer };
+            case CSS::ValueID::RubyTextContainer:
+                return CSS::Display { CSS::Display::Internal::RubyTextContainer };
+
+            // display-box
+            case CSS::ValueID::Contents:
+                // FIXME this should be CSS::Display::Box::Contents but contents is currently not implemented
+                return CSS::Display { CSS::Display::Outside::Block, CSS::Display::Inside::Flow };
+            case CSS::ValueID::None:
+                return CSS::Display { CSS::Display::Box::None };
+
+            // display-legacy
+            case CSS::ValueID::InlineBlock:
+                return CSS::Display { CSS::Display::Outside::Inline, CSS::Display::Inside::FlowRoot };
+            case CSS::ValueID::InlineTable:
+                return CSS::Display { CSS::Display::Outside::Inline, CSS::Display::Inside::Table };
+            case CSS::ValueID::InlineFlex:
+                return CSS::Display { CSS::Display::Outside::Inline, CSS::Display::Inside::Flex };
+            case CSS::ValueID::InlineGrid:
+                return CSS::Display { CSS::Display::Outside::Inline, CSS::Display::Inside::Grid };
+
+            default:
+                return {};
+            }
+        }
+        return {};
+    };
+
+    auto parse_double_component_display = [&](Vector<ComponentValue> const& component_values) -> Optional<CSS::Display> {
+        auto first_identifier = parse_identifier_value(component_values.first());
+        auto second_identifier = parse_identifier_value(component_values[1]);
+        if (first_identifier && second_identifier)
+            if (is_display_outside(first_identifier->to_identifier()) && is_display_inside(second_identifier->to_identifier()))
+                return CSS::Display { parse_outside(first_identifier->to_identifier()), parse_inside(second_identifier->to_identifier()) };
+
+        if (CSS::ValueID::ListItem == second_identifier->to_identifier()) {
+            if (is_display_outside(first_identifier->to_identifier()))
```

Comment:
Same trap here. `<display-listitem> = <display-outside>? && [ flow | flow-root ]? && list-item` might have the three parts in any order.

There's not an easy way to parse things like this. My usual approach (which you can find in a few places in the parser, for example `parse_list_style_value()`) is to loop through the component values and check each one to see which "slot" it can fill.

Reply:
What do you think should we do with a value like `block table list-item`? This is illegal when following the spec so should we just raise an SyntaxError?

Reply:
I implemented a "slot" system so it should be better now, i could even reduce the amount of code. I now treat something like `block table list-item` as a SyntaxError.